tests/fuser_basic_test.cc

```diff
@@ -1104,6 +1108,336 @@ TEST(FuserBasicTest, ControlledGate) {
+      GateCZ<float>::Create(1, 1, 2),
+      GateCZ<float>::Create(1, 3, 4),
+      GateCZ<float>::Create(0, 6, 7),
+      GateCZ<float>::Create(1, 5, 6),
+      GateCZ<float>::Create(1, 7, 0),
+    };
+
+    auto fused_gates = Fuser::FuseGates(
+        param, num_qubits, circuit.begin(), circuit.end());
+
+    EXPECT_EQ(fused_gates.size(), 8);
+    EXPECT_TRUE(TestFusedGates(num_qubits, circuit, fused_gates));
+  }
+
+  {
+    unsigned num_qubits = 4;
+    std::vector<Gate> circuit = {
+      GateCZ<float>::Create(0, 0, 1),
+      GateCZ<float>::Create(0, 2, 3),
+      GateCZ<float>::Create(2, 1, 2),
+      GateCZ<float>::Create(1, 0, 3),
+      GateCZ<float>::Create(3, 1, 2),
+      GateCZ<float>::Create(3, 0, 3),
+      GateCZ<float>::Create(5, 1, 2),
+      GateCZ<float>::Create(4, 0, 3),
+    };
+
+    std::vector<unsigned> time_boudary = {3};
+    auto fused_gates = Fuser::FuseGates(
+        param, num_qubits, circuit.begin(), circuit.end(), time_boudary);
+
+    EXPECT_EQ(fused_gates.size(), 6);
+    EXPECT_TRUE(TestFusedGates(num_qubits, circuit, fused_gates));
+  }
+
+  {
+    unsigned num_qubits = 4;
+    std::vector<Gate> circuit = {
+      GateCZ<float>::Create(0, 0, 1),
+      GateCZ<float>::Create(0, 2, 3),
+      GateCZ<float>::Create(2, 1, 2),
+      GateCZ<float>::Create(1, 0, 3),
+      gate::Measurement<Gate>::Create(3, {1, 2}),
+      GateCZ<float>::Create(3, 0, 3),
+      GateCZ<float>::Create(5, 1, 2),
+      GateCZ<float>::Create(4, 0, 3),
+    };
+
+    auto fused_gates = Fuser::FuseGates(
+        param, num_qubits, circuit.begin(), circuit.end());
+
+    EXPECT_EQ(fused_gates.size(), 7);
+    EXPECT_TRUE(TestFusedGates(num_qubits, circuit, fused_gates));
+  }
+}
+
+TEST(FuserBasicTest, InvalidTimeOrder) {
+  using Gate = GateQSim<float>;
+  using Fuser = BasicGateFuser<IO, Gate>;
+
+  Fuser::Parameter param;
+  param.verbosity = 0;
+
+  {
+    unsigned num_qubits = 3;
+    std::vector<Gate> circuit = {
+      GateCZ<float>::Create(0, 0, 1),
+      GateCZ<float>::Create(0, 1, 2),
+    };
+
+    auto fused_gates = Fuser::FuseGates(
+        param, num_qubits, circuit.begin(), circuit.end());
+
+    EXPECT_EQ(fused_gates.size(), 0);
+  }
+
+  {
+    unsigned num_qubits = 4;
+    std::vector<Gate> circuit = {
+      GateCZ<float>::Create(0, 0, 1),
+      GateCZ<float>::Create(0, 2, 3),
+      GateCZ<float>::Create(2, 1, 2),
+      GateCZ<float>::Create(1, 0, 2),
+    };
+
+    auto fused_gates = Fuser::FuseGates(
+        param, num_qubits, circuit.begin(), circuit.end());
+
+    EXPECT_EQ(fused_gates.size(), 0);
+  }
+
+  {
+    unsigned num_qubits = 4;
+    std::vector<Gate> circuit = {
+      GateCZ<float>::Create(0, 0, 1),
+      GateCZ<float>::Create(0, 2, 3),
+      GateCZ<float>::Create(2, 0, 3),
+      GateCZ<float>::Create(1, 1, 2),
+    };
+
+    std::vector<unsigned> time_boudary = {1};
+    auto fused_gates = Fuser::FuseGates(
+        param, num_qubits, circuit.begin(), circuit.end(), time_boudary);
```

Comment:
`time_boudary` -> `time_boundary`

Same elsewhere in file.

Reply:
Done.